dll/win32/shell32/utils.cpp

```diff
@@ -31,6 +31,316 @@ OpenEffectiveToken(
+HRESULT
+Shell_DisplayNameOf(
+    _In_ IShellFolder *psf,
+    _In_ LPCITEMIDLIST pidl,
+    _In_ DWORD dwFlags,
+    _Out_ LPWSTR pszBuf,
+    _In_ UINT cchBuf)
+{
+    *pszBuf = UNICODE_NULL;
+    STRRET sr;
+    HRESULT hr = psf->GetDisplayNameOf(pidl, dwFlags, &sr);
+    if (FAILED(hr))
+        return hr;
+    return StrRetToBufW(&sr, pidl, pszBuf, cchBuf);
+}
+
+DWORD
+SHGetAttributes(_In_ IShellFolder *psf, _In_ LPCITEMIDLIST pidl, _In_ DWORD dwAttributes)
+{
+    LPCITEMIDLIST pidlLast;
+
+    if (psf)
+    {
+        psf->AddRef();
+        pidlLast = pidl;
+    }
+    else
+    {
+        SHBindToParent(pidl, IID_PPV_ARG(IShellFolder, &psf), &pidlLast);
+    }
+
+    if (!psf)
+        return 0;
+
+    DWORD oldAttrs = dwAttributes;
+    if (FAILED(psf->GetAttributesOf(1, &pidlLast, &dwAttributes)))
+        dwAttributes = 0;
+    else
+        dwAttributes &= oldAttrs;
+
+    if ((dwAttributes & SFGAO_FOLDER) &&
+        (dwAttributes & SFGAO_STREAM) &&
+        !(dwAttributes & SFGAO_STORAGEANCESTOR) &&
+        (oldAttrs & SFGAO_STORAGEANCESTOR) &&
+        (SHGetObjectCompatFlags(psf, NULL) & 0x200))
+    {
+        dwAttributes &= ~(SFGAO_STREAM | SFGAO_STORAGEANCESTOR);
+        dwAttributes |= SFGAO_STORAGEANCESTOR;
+    }
+
+    if (psf)
+        psf->Release();
+
+    return dwAttributes;
+}
+
+HRESULT SHCoInitializeAnyApartment(VOID)
+{
+    HRESULT hr = CoInitializeEx(NULL, COINIT_APARTMENTTHREADED | COINIT_DISABLE_OLE1DDE);
+    if (FAILED(hr))
+        hr = CoInitializeEx(NULL, COINIT_DISABLE_OLE1DDE);
+    return hr;
+}
+
+HRESULT
+SHGetNameAndFlagsW(
+    _In_ LPCITEMIDLIST pidl,
+    _In_ DWORD dwFlags,
+    _Out_opt_ LPWSTR pszText,
+    _In_ UINT cchBuf,
+    _Inout_opt_ DWORD *pdwAttributes)
+{
+    if (pszText)
+        *pszText = UNICODE_NULL;
+
+    HRESULT hrCoInit = SHCoInitializeAnyApartment();
+
+    CComPtr<IShellFolder> psfFolder;
+    LPCITEMIDLIST ppidlLast;
+    HRESULT hr = SHBindToParent(pidl, IID_PPV_ARG(IShellFolder, &psfFolder), &ppidlLast);
+    if (SUCCEEDED(hr))
+    {
+        if (pszText)
+            hr = Shell_DisplayNameOf(psfFolder, ppidlLast, dwFlags, pszText, cchBuf);
+
+        if (SUCCEEDED(hr))
+        {
+            if (pdwAttributes)
+                *pdwAttributes = SHGetAttributes(psfFolder, ppidlLast, *pdwAttributes);
+        }
+    }
+
+    if (SUCCEEDED(hrCoInit))
+        CoUninitialize();
+
+    return hr;
+}
+
+EXTERN_C HWND
+BindCtx_GetUIWindow(_In_ IBindCtx *pBindCtx)
+{
+    HWND hWnd = NULL;
+
+    CComPtr<IUnknown> punk;
+    if (pBindCtx && SUCCEEDED(pBindCtx->GetObjectParam((LPWSTR)L"UI During Binding", &punk)))
+        IUnknown_GetWindow(punk, &hWnd);
+
+    return hWnd;
+}
+
+class CDummyOleWindow : public IOleWindow
+{
+protected:
+    LONG m_cRefs;
+    HWND m_hWnd;
+
+public:
+    CDummyOleWindow() : m_cRefs(1), m_hWnd(NULL) { }
+    virtual ~CDummyOleWindow() { }
+
+    // IUnknown methods
+    STDMETHODIMP QueryInterface(REFIID riid, LPVOID *ppvObj) override
```

Comment:
Would you consider ommitting the overrides? It breaks compatiblity to older compilers.

Reply:
Those compilers (before C++11) are too oldie. The `override` keyword verifies overriding.
That is important.

Reply:
I do have a different opinion on this, therefore I will not approve this PR (but will also not actively reject it).

Reply:
https://jira.reactos.org/browse/CORE-19509

Reply:
@JoachimHenze can you stop beating this already mummified dead horse?
It is pretty annoying to see you commenting on multiple PR's, asking people to write worse code to support something that we decided multiple years ago not to support anymore.